src/main/java/com/iota/iri/hash/Kerl.java

```diff
@@ -6,142 +6,174 @@
+            e.printStackTrace(System.err);
             throw new RuntimeException(e);
         }
     }
 
     public static BigInteger bigIntFromTrits(final int[] trits, final int offset, final int size) {
-
+        for (int i = offset; i < offset + size; i++) {
+            if (trits[i] < -1 || trits[i] > 1) {
+                throw new IllegalArgumentException("not a trit: " + trits[i]);
+            }
+        }
         BigInteger value = BigInteger.ZERO;
-
-        for (int i = size; i-- > 0; ) {
-
-            value = value.multiply(BigInteger.valueOf(Converter.RADIX)).add(BigInteger.valueOf(trits[offset + i]));
+        long num;
+        int power = 0;
+        int n = offset + size - 1;
+        for (; n > offset + 1; n--) {
+            switch (trits[n]) {
+                case 0: // a * b
+                    power++;
+                    break;
+                case 1: // a * b + 1
+                    num = 9 + (trits[--n] * 3) + trits[--n];
+                    for (int count = 0; n > offset && count <= 36; count++, power++) {
+                        num = 3 * num + trits[--n];
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 3]).add(BigInteger.valueOf(num));
+                    power = 0;
+                    break;
+                case -1: // a * b - 1
+                    num = 9 - (trits[--n] * 3) - trits[--n];
+                    for (int count = 0; n > offset && count <= 36; count++, power++) {
+                        num = 3 * num - trits[--n];
+                    }
+                    value = value.multiply(RADIX_POWERS[power + 3]).subtract(BigInteger.valueOf(num));
+                    power = 0;
+                    break;
+
+                default:
+                    throw new RuntimeException("unreachable");
+            }
+        }
+        for (; n >= offset; n--) {
+            switch (trits[n]) {
+                case 0:
+                    power++;
+                    break;
+                case 1:
+                    value = value.multiply(RADIX_POWERS[power + 1]).add(BigInteger.ONE);
+                    power = 0;
+                    break;
+                case -1:
+                    value = value.multiply(RADIX_POWERS[power + 1]).subtract(BigInteger.ONE);
+                    power = 0;
+                    break;
+                default:
+                    throw new RuntimeException("unreachable");
+            }
+        }
+        // do leftover pow
+        if (power > 0) {
+            value = value.multiply(RADIX_POWERS[power]);
         }
-
         return value;
     }
 
     public static BigInteger bigIntFromBytes(final byte[] bytes, final int offset, final int size) {
-
-        return new BigInteger(Arrays.copyOfRange(bytes, offset, offset + size));
+        return (bytes.length == offset + size) ? new BigInteger(bytes)
+            : new BigInteger(Arrays.copyOfRange(bytes, offset, offset + size));
     }
 
```

Comment:
bigIntFromBytes - most times array is proper size and no copy is needed. Added conditional copy.